File: 1792-maximum-average-pass-ratio/1792-maximum-average-pass-ratio.cpp
class Solution {
public:
    double maxAverageRatio(vector<vector<int>>& classes, int extraStudents) {
        priority_queue<pair<double, int>> heap;
        int i = 0;
        for (auto cls: classes){
            int passing = cls[0];
            int total = cls[1];
            double current = (double)(passing) / total;
            double neww = (double)(passing + 1) / (total + 1); 
            heap.push({neww - current, i});
            i += 1;
        }

        for (int i = 0; i < extraStudents; i++){
            auto [diff, idx] = heap.top();
            heap.pop();
            classes[idx][0] += 1;
            classes[idx][1] += 1;
            int passing = classes[idx][0];
            int total = classes[idx][1];
            double current = (double)(passing) / total;
            double neww = (double)(passing + 1) / (total + 1); 
            heap.push({neww - current, idx});
        }

        long double total = 0;
        for (auto cls: classes){
            total += (double)(cls[0]) / cls[1];
        }

        return total / classes.size();
    }
};

File: 1792-maximum-average-pass-ratio/1792-maximum-average-pass-ratio.py
class Solution:
    def maxAverageRatio(self, classes: List[List[int]], extraStudents: int) -> float:
        heap = []
        for i in range(len(classes)):
            passing, total = classes[i]
            curr = passing / total
            new = (passing + 1) / (total + 1)
            heappush(heap, (-(new - curr), i))
        
        for _ in range(extraStudents):
            _, i = heappop(heap)
            classes[i] = [classes[i][0] + 1, classes[i][1] + 1]
            curr = classes[i][0] / classes[i][1]
            new = (classes[i][0] + 1) / (classes[i][1] + 1)
            heappush(heap, (-(new - curr), i))
        
        tot = 0
        for passing, total in classes:
            tot += passing / total
        
        return tot / len(classes)
            